Improve startup experience by removing AppBar skeleton

LasseRosenow opened the pull request:
## Proposed change

Currently when launching the Home Assistant Web App, there appears a small AppBar skeleton on mobile for a very short time, disappers again and then the real layout is loaded in.
This looks very strange to the eye, so I figured it would be much cleaner to remove this skeleton and have a more clean transition.

Also its confusing anyways to have a skeleton that does not follow the seme of the app, that then is later replaced with a Themed AppBar that might look totally different.

### How it is
<table>
<tr><th>AppBar</th><th>No AppBar</th><th>Themed AppBar</th></tr>
<tr>
<td>

![1](https://user-images.githubusercontent.com/10547444/140663948-2383d7fa-16c2-4cc7-ae92-3f7dd50d8080.png)

</td>
<td>

![2](https://user-images.githubusercontent.com/10547444/140663980-5d0fb087-c5d3-46b5-9384-ce894a73f3b9.png)

</td>
<td>

![3](https://user-images.githubusercontent.com/10547444/140664063-d9f1fda2-9028-478f-94d0-d5e68bbcafc6.png)

</td>
</tr>
</table>

### How it should be
<table>
<tr><th>No AppBar</th><th>No AppBar</th><th>Themed AppBar</th></tr>
<tr>
<td>

![1 new](https://user-images.githubusercontent.com/10547444/140663988-773dcad9-c9dc-4ef6-8584-f6b6bb9510ca.png)

</td>
<td>

![2](https://user-images.githubusercontent.com/10547444/140663980-5d0fb087-c5d3-46b5-9384-ce894a73f3b9.png)

</td>
<td>

![3](https://user-images.githubusercontent.com/10547444/140664063-d9f1fda2-9028-478f-94d0-d5e68bbcafc6.png)

</td>
</tr>
</table>

Especially in the android app this would get even weirder, as you would have this order of screens:
1. blank android launch activity
2. AppBar skeleton
3. Loading screen without AppBar
4. Themed AppBar

So it would be -- off -- on -- off -- on --

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

LasseRosenow committed:
Remove appbar skeleton to fix startup blinking

LasseRosenow commented:
Here is also a video demonstrating the issue on android

https://user-images.githubusercontent.com/10547444/140664336-58918b45-d7df-45ec-a3d0-5249afa1b922.mp4

bramkragten commented:
Is your device in dark mode?

LasseRosenow commented:
Yes, I posted the screenshots in light mode, so you could see the issue more clearly, but it is also a problem if the app is in dark mode too, as you would still get a subtle flashing of na app bar that appears and then disappears again. Just then to appear again.

I can also add screenshots with the app in dark mode too later on to explain further what I mean if you would like that :)

bramkragten commented:
It explains why the app bar is black in the first image, as if your device would be in light mode that app bar would be blue.

I think the real issue here is that the app bar is missing on the `ha-init-page`...?

bramkragten commented on src/html/index.html.template:
```diff
@@ -53,14 +47,10 @@
           color: #e1e1e1;
         }
-        #ha-init-skeleton::before {
-          background-color: #1c1c1c;
```
This color should be updated, as it is wrong now: https://github.com/home-assistant/frontend/blob/dev/src/resources/styles.ts#L11

LasseRosenow commented:
I just checked, but the AppBar is not blue in light mode. This seems to be because the `#THEMEC` variable is replaced by `var(--primary-background-color)`

![image](https://user-images.githubusercontent.com/10547444/140733256-ad71aa7f-d6c4-4cd9-9267-fcdfdcf1675a.png)

But this variable seems not to be set yet, so its transparent

bramkragten commented:
That would be your theme, it uses the `app-header-background-color` from your default theme

LasseRosenow commented:
I switched to the default home assistant theme, but its not blue 🤔

bramkragten commented:
![image](https://user-images.githubusercontent.com/5662298/140734447-b9b16da7-06e1-4918-b37a-69fa791cc1dc.png)

LasseRosenow commented:
I see you are right. Very interesting. But then every user that uses a different color would end up having this wrong title bar color on startup. Many Home Assistant users don't even know how to set up a default backend theme and they would just set the theme for the user and thats it.

I think removing this would still be the best way, as for example in my case my custom theme has a white AppBar color and if I would not have set is as the default backend theme, my startup sequence would be to have a blue AppBar first, that then would be replaced by a white one later on. Which looks not ideal in my believe.

In my believe for applications like Home Assistant, that are very customizable etc. it would be best just to draw simple splash screens without any theme specific ui that would then be overriden by a dynamic theme that is loaded in later on.

balloob commented:
The app skeleton is drawn so a user knows something is happening. If you just see a blank screen, you might not be aware that anything is loading and think it's broken. Especially on a slower connection and/or older phones it might be a while before the loading spinner shows. For that reasonn, removing it without adding an alternative is not an option.

LasseRosenow commented:
I see your point.

I still think that everything that is bound to a theme is not a good option, so lets think about good replacements.

My Idea would be to show this icon:
![image](https://user-images.githubusercontent.com/10547444/140800665-d5cece38-2d36-4992-be6e-fd32322e484b.png)

But as a svg of course, so it would load instantly.. (could also be inline)

And then to prevent a weird transition from this logo to the loading screen, maybe remove the loading circle from the loading screen and just show only this logo as well.

This would result in a very clean startup on most of the devices, as the loading circle is not helping if it is only shown for 1s, which would be the case an most devices. And it would also give some information for devices with bad internet, that stuff is happening.

To further improve this, I would propose to start show the loading circle, if loading takes more then 5s (or any other value that would make sense)

So in the end we would have:

1. blank screen with Logo (html page)
2. blank screen with Logo (ha-init-page)

if loading takes longer than 5s:
3. blank screen with logo and circle below

balloob commented:
That's a great proposal. Note that `<ha-init-page>` also includes the text "Loading Data" below the spinner. I think that we should keep that in place so you can see some text showing indicating something is happening. 

For spinner -> let's also load it on demand then, instead of importing it at the top, that should help reduce initial bundle.

LasseRosenow committed:
Implement new launch screen

LasseRosenow commented:
So I just implemented the new launch screen.

I got the logo svg from homeassistant.io and added a border radius to match the one from the old logo.
I changed the html template so that it contains the logo instead of the app bar skeleton. Then I changed the `ha-init-page` so that it does not contain the home assistant logo any more to prevent code duplication and future inconsistencies. What I'm doing now is injecting the `ha-init-page` into the launch screen instead of hiding the launch screen from the html template and drawing on top of it.

Here are the results:

<table>
<tr><th>Pure HTML Launch Screen</th><th>ha-init-page injected</th><th>After 5 seconds</th></tr>
<tr>
<td>

![localhost_8090_(iPhone X)](https://user-images.githubusercontent.com/10547444/141667379-77e38e68-1930-4b7c-831d-c56c543d06f3.png)


</td>
<td>

![localhost_8090_(iPhone X) (1)](https://user-images.githubusercontent.com/10547444/141667382-5fabffe1-42c9-4629-9dfb-c1d5313cb7e5.png)


</td>
<td>


![localhost_8090_(iPhone X) (2)](https://user-images.githubusercontent.com/10547444/141667385-5fddf219-dac6-4d5c-99ed-e50a0d05d979.png)

</td>
</tr>
</table>

balloob commented:
It looks like "loading data" jumps to the right when the spinner appears. Can we make the spinner sit above loading data instead?

balloob commented:
This is really awesome btw! <3

balloob commented on src/layouts/ha-init-page.ts:
```diff
@@ -1,43 +1,43 @@
 
   protected firstUpdated() {
-    removeInitSkeleton();
+    setTimeout(async () => {
```
Let's clear this timeout when this element is disconnected.

balloob commented on src/layouts/home-assistant.ts:
```diff
@@ -57,14 +61,20 @@ export class HomeAssistantAppEl extends QuickBarMixin(HassElement) {
-        `
-      : html`<ha-init-page .error=${this._error}></ha-init-page>`;
+    if (hass && hass.states && hass.config && hass.services) {
+      removeLaunchScreen();
```
We should not do this on every render. `render()` cannot have side effects.

Makes me wonder if we should drop `LitElement` here and extend `ReactiveElement` directly so we can implement `update()` (which is meant to have side effects)

LasseRosenow replied:
LitElement actually allows to implement update() so that should be an easy fix

balloob replied:
LitElement = Reactive Element with an update implementation being kinda like `render(this.shadowRoot, this.render())`. Looked it up just now, I was spot on :-) https://github.com/lit/lit/blob/main/packages/lit-element/src/lit-element.ts#L136

We shouldn't override `update()` from LitElement, but instead implement it for ReactiveElement.

LasseRosenow replied:
But we still need the render() method to from LitElement to render `<home-assistant-main>`

I think the usual way to do this would be to implement update() like this:
```typescript

  update(changedProps) {
    super.update(changedProps);

    const hass = this.hass;
    if (this._isHassComplete()) {
      removeLaunchScreen();
    }
  }

```

LasseRosenow commented:
Thank you very much :)

Putting the spinner above the text would result in the text jumping down instead of jumping right or there would be quite a huge gap between the logo and the text until the spinner appears.

I was also thinking about maybe just animating the text to float right and fading in the spinner 🤔 

So the possible solutions I could think of would be:

1. Loading circle pushes the text to the right (with / without animation)
2. Loading circle pushes the text down (with / without animation)
3. There is a initial gap between the logo and the text so the loading circle can just appear and fill it up without pushing anything around.

Each of these solutions would be very easy and fast for me to implement and I don't have any strong opinion on which would be best, so just pick whatever you think would be best :)

But before you decide I would like to express, that it is more important to have the screen without the loading circle look good than the screen with it, as seeing the loading screen should be quite rare for most of the users I suppose.

balloob commented:
Let's go for initial gap.

LasseRosenow committed:
Address feedback

LasseRosenow commented:
So I just updated the code as per your feedback. This is how it looks like:

<table>
<tr><th>Pure HTML Launch Screen</th><th>ha-init-page injected</th><th>After 5 seconds</th></tr>
<tr>
<td>

![localhost_8090_(iPhone X)](https://user-images.githubusercontent.com/10547444/141667379-77e38e68-1930-4b7c-831d-c56c543d06f3.png)


</td>
<td>

![localhost_8090_(iPhone X) (5)](https://user-images.githubusercontent.com/10547444/141686217-3e7df72f-c0e4-46a1-81c7-13d1bafcd46b.png)



</td>
<td>


![localhost_8090_(iPhone X) (6)](https://user-images.githubusercontent.com/10547444/141686219-9584b9ab-8184-4517-8ff2-af4baa1149b3.png)


</td>
</tr>
</table>

balloob commented:
I commented out `import("/frontend_latest/app.js");` in `hass_frontend/index.html` to avoid the app from loading and noticed that the logo is top aligned:

![image](https://user-images.githubusercontent.com/1444314/141731431-60d94367-765f-4bfc-905f-01084e6e7a22.png)

This is on the latest commit 25ab7fbf4d22cc8b7b995b1415482e4fb064bde1

balloob commented:
Oh another cool idea, should we have the logo Css animate in? Small playful animation

LasseRosenow commented:
Uh that's strange, I'll try to reproduce this issue later.


I think animating the logo in would be a bit overkill, also it would clash with the android and iOS counter parts as my idea was to have those app use the logo in their launch screen aswell. This would result in having a logo, then not having a logo and then seeing the same logo fading in

LasseRosenow committed:
Fix launch screen position

LasseRosenow commented:
So the issue with the position of the logo should be fixed now

balloob commented:
This is perfect, thank you so much! ❤️
